lib/symlink/index.js

```diff
@@ -51,22 +51,24 @@ function symlink(outFolder, opt) {
       srcPath = path.relative(file.base, srcPath);
     }
 
-    // TODO: make DRY with .dest()
-    var dirMode = number(opt.dirMode, file);
-    var writeFolder = path.dirname(file.path);
-
-    fo.mkdirp(writeFolder, dirMode, onMkdirp);
+    // Because fs.symlink does not allow atomic overwrite option with flags, we
+    // delete and recreate if the link already exists and overwrite is true.
+    if (file.flag === 'w') {
+      // TODO What happens when we call unlink with windows junctions?
+      fs.unlink(file.path, onUnlink);
+    } else {
+      fs.symlink(srcPath, file.path, symType, onSymlink);
+    }
 
-    function onMkdirp(mkdirpErr) {
-      if (mkdirpErr) {
-        return callback(mkdirpErr);
+    function onUnlink(unlinkErr) {
+      if (unlinkErr && unlinkErr.code !== 'ENOENT') {
+        return callback(unlinkErr);
       }
-
       fs.symlink(srcPath, file.path, symType, onSymlink);
     }
 
     function onSymlink(symlinkErr) {
-      if (isErrorFatal(symlinkErr)) {
+      if (fo.isOverwriteError(symlinkErr, file.flag)) {
```

Comment:
Sorry, I guess that wasn't a very clear description on my part. So, the code for `symlinks` previously did not look at the `overwrite` option (or its derivative `file.flag` property) at all, and always assumed that an `EEXIST` error from `fs.symlink`could be ignored (as if `overwrite` were false).

The tests for `symlink` did not consider any overwrite scenarios (never attempted to symlink at an existing path) and so this never popped up.

So in this PR I've tried to account for `overwrite` with symlinks, and to make this consistent with how `dest` deals with it. My remark about having copied tests (almost) verbatim from `dest` was just to reinforce the last bit, that it now behaves (or it should) consistent with `dest`.